pkg/hubble/getters.go

```diff
@@ -0,0 +1,162 @@
+	return &LocalIdentityGetter{allocator: allocator}
+}
+
+// GetIdentity looks up identity by ID from Cilium's identity cache. Hubble uses the identity info
+// to populate source and destination labels of flows.
+func (getter *LocalIdentityGetter) GetIdentity(securityIdentity uint64) (*models.Identity, error) {
+	ident := getter.allocator.LookupIdentityByID(context.Background(), identity.NumericIdentity(securityIdentity))
+	if ident == nil {
+		return nil, fmt.Errorf("identity %d not found", securityIdentity)
+	}
+	return ident.GetModel(), nil
+}
+
+// LocalEndpointDNSGetter implements DNSGetter and EndpointGetter interfaces from Hubble.
+//  - EndpointGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L34
+//  - DNSGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L27
+type LocalEndpointDNSGetter struct {
+	manager *endpointmanager.EndpointManager
+}
+
+// NewLocalEndpointDNSGetter returns an initialized pointer to LocalEndpointDNSGetter.
+func NewLocalEndpointDNSGetter(manager *endpointmanager.EndpointManager) *LocalEndpointDNSGetter {
+	return &LocalEndpointDNSGetter{manager: manager}
+}
+
+// GetEndpoint returns endpoint info for a given IP address. Hubble uses this function to populate
+// fields like namespace and pod name for local endpoints.
+func (getter *LocalEndpointDNSGetter) GetEndpoint(ip net.IP) (endpoint *hubbleV1.Endpoint, ok bool) {
+	ep := getter.manager.LookupIP(ip)
+	if ep == nil {
+		return nil, false
+	}
+	return hubbleEndpoint.ParseEndpointFromModel(ep.GetModel()), true
+}
+
+// GetNamesOf implements DNSGetter.GetNamesOf. It looks up DNS names of a given IP from the
+// FQDN cache of an endpoint specified by sourceEpID.
+func (getter *LocalEndpointDNSGetter) GetNamesOf(sourceEpID uint64, ip net.IP) []string {
+	ep := getter.manager.LookupCiliumID(uint16(sourceEpID))
+	if ep == nil {
+		return nil
+	}
+	return ep.DNSHistory.LookupIP(ip)
+}
+
+// LocalIPGetter implements IPGetter interface from Hubble.
+//  - IPGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L46
+type LocalIPGetter struct {
+	ipCache *ipcache.IPCache
+}
+
+// NewLocalIPGetter returns an initialized pointer to LocalIPGetter.
+func NewLocalIPGetter(ipCache *ipcache.IPCache) *LocalIPGetter {
+	return &LocalIPGetter{ipCache: ipCache}
+}
+
+// GetIPIdentity returns the IP identity of the given IP address. Hubble uses this function to populate
+// fields like namespace and pod name for remote endpoints.
+func (getter *LocalIPGetter) GetIPIdentity(ip net.IP) (hubbleIPCache.IPIdentity, bool) {
+	ipIdentity, ok := getter.ipCache.LookupByIP(ip.String())
+	if !ok {
+		return hubbleIPCache.IPIdentity{}, false
+	}
+	meta := getter.ipCache.GetK8sMetadata(ip.String())
+	if meta == nil {
+		return hubbleIPCache.IPIdentity{}, false
```

Comment:
Not all identity must have K8s Metadata, e.g. CIDR identities are fine to just have the `Identity` field populated. We want to return 
```go
return hubbleIPCache.IPIdentity{
    Identity: ipIdentity.ID,
}, true
```
here.